sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala

```diff
@@ -229,3 +208,27 @@ class EquivalentExpressions {
     sb.toString()
   }
 }
+
+/**
+ * Orders `Expression` by parent/child relations. The child expression is smaller
+ * than parent expression. If there is child-parent relationships among the subexpressions,
+ * we want the child expressions come first than parent expressions, so we can replace
+ * child expressions in parent expressions with subexpression evaluation. Note that
+ * this is not for general expression ordering. For example, two irrelevant expressions
+ * will be considered as e1 < e2 and e2 < e1 by this ordering. But for the usage here,
+ * the order of irrelevant expressions does not matter.
+ */
+class ExpressionContainmentOrdering extends Ordering[Expression] {
```

Comment:
Just curious, is there a reason of this move?

Reply:
Oh, as it is a nested class, I cannot allocate it separately, but

```scala
val equivalence = new EquivalentExpressions
val exprOrdering = new equivalence.ExpressionContainmentOrdering
```

I can revert to nested class if you think it's unnecessary change.

Reply:
Never mind. New one also looks good~